lib/Doctrine/Enumeration/Enumerated.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Enumeration;
+
+use Doctrine\Enumeration\Exception\NotSupportedException;
+use function array_search;
+use function constant;
+use function in_array;
+
+trait Enumerated
+{
+    final private function __construct()
+    {
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function validateValue($value) : void
+    {
+        if (! in_array($value, self::availableValues(), true)) {
+            throw new NotSupportedException('Invalid value.');
+        }
+    }
+
+    final public static function validateName(string $name) : void
+    {
+        if (! isset(self::availableValues()[$name])) {
+            throw new NotSupportedException('Invalid name.');
+        }
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function get($value) : self
+    {
+        self::validateValue($value);
+
+        $name      = array_search($value, self::availableValues(), true);
+        $instances = &self::registry(static::class);
+
+        return $instances[$name] ?? $instances[$name] = new static();
+    }
+
+    /**
+     * @return mixed
+     */
+    final private static function value(self $instance)
+    {
+        return self::availableValues()[array_search($instance, self::registry(static::class), true)];
+    }
+
+    /**
+     * @return self[]
+     */
+    private static function &registry(string $type) : array
+    {
+        static $registry = [];
+
+        if (! isset($registry[$type])) {
+            $registry[$type] = [];
+        }
+
+        $ref = &$registry[$type];
+        return $ref;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    private static function availableValues() : array
+    {
+        static $values = [];
+
+        if (isset($values[static::class])) {
+            return $values[static::class];
+        }
+
+        $values[static::class] = [];
+
+        foreach ((new \ReflectionClass(static::class))->getReflectionConstants() as $name => $constantReflection) {
```

Comment:
That looks like uber magic. I thought the limited number of existing objects will be enforced by the fact that the class creates its own instances and stores them indexed by name. Not like this.

Reply:
This is actually the correct way to do it, although we should probably override `availableValues()` in userland where speed may be a problem (usually not though, since this is only ever executed once.

Reply:
Could we take a simpler approach without reflection and array lookups?
```php
class Enum
{
    private static $instances;

    private function __construct()
    {
    }

    public static function X() : self
    {
        return self::value(__FUNCTION__);
    }

    public static function Y() : self
    {
        return self::value(__FUNCTION__);
    }

    private static function value(string $name) : self
    {
        return self::$instances[$name] ?? self::$instances[$name] = new self();
    }
}

var_dump(Enum::X() === Enum::X());
// true

var_dump(Enum::Y() === Enum::Y());
// true

var_dump(Enum::X() === Enum::Y());
// false
```

Reply:
Your proposal is no longer by value and changes the semantics. Also it wouldn't be possible to use `$fooBar() === $foo::BAR`.

Reply:
Why expose and even constants? Doesn't it break encapsulation in the way that we want to operate objects, so their associated scalar values are just implementation details.

Why cannot the caller do `$fooBar() === Foo::BAR()` instead?

Reply:
If it's because of BC, I think we need to decide first what the proper usage should look like and then build the compatibility layer. Mixing things together looks confusing from the design standpoint (unclear what's by design and what's temporary BC layer).

Reply:
BC is only one part. It's also because the code with constants is a bit more comprehensible. Without those, everything is anonymous and you need magic like `__FUNCTION__` or so. Also instance can't be used in arrays - that is often useful for maps (i.e. ParameterType mappings in drivers).

Reply:
I'm not saying it's not possible or that it's a bad idea. But in today's PHP I think using constants gives more flexibility, especially allows easily building maps of enums (their values) to mapped values by simply using the constant and then dereferencing the enum using `()`.

Reply:
I feel lost. Or, to put it better, I understand the reasoning but I find it confusing that plain constants get replaced with a relatively complex concept to be able to use object and type-hint against their classes, but still, we want to use constants as well.

The fact that the same thing can be referred to by multiple means, and the connection between them is hidden somewhere in the framework, looks similar to the magic which people hate about frameworks like Laravel. This approach may be usable from the consumer's standpoint once you get the concept — like Laravel's facades (which I never used) — but it just doesn't feel straightforward.

Reply:
Ok, consider we have this currently (in OCI8Statement):
```php
    protected static $fetchModeMap = [
        FetchMode::MIXED       => OCI_BOTH,
        FetchMode::ASSOCIATIVE => OCI_ASSOC,
        FetchMode::NUMERIC     => OCI_NUM,
        FetchMode::COLUMN      => OCI_NUM,
    ];
```

How would you propose to rewrite this with the completely anonymous approach?
This is invalid:
```php
    protected static $fetchModeMap = [
        FetchMode::MIXED()       => OCI_BOTH,
        FetchMode::ASSOCIATIVE() => OCI_ASSOC,
        FetchMode::NUMERIC()     => OCI_NUM,
        FetchMode::COLUMN()      => OCI_NUM,
    ];
```

You would need to move this into a private method, build a SplObjectStorage map in constructor or something like that.

Reply:
To me, a replacement of the map with a block of `if`/`else` would be more acceptable.

Reply:
Ok, I will think about it. 👍
@ocramius WDYT?

Reply:
The reflection approach is good enough for me. It is an internal detail that is an `O(1)`, and no I/O is involved, so there is no noticeable overhead. Being able to override the "default magicness" is as easy as importing the method from the trait with a different name